pact/exchange.pact

```diff
@@ -0,0 +1,582 @@
+(enforce-pact-version "3.7")
+
+(namespace 'swap)
+
+(module exchange GOVERNANCE
+
+  (defcap GOVERNANCE () (enforce false "autonomous"))
+
+  (defcap CREATE_PAIR
+    ( token0:module{fungible-v2}
+      token1:module{fungible-v2}
+      key:string
+      account:string )
+    " Pair-created event for TOKEN0 and TOKEN1 pairs with KEY liquidity token \
+    \ and ACCOUNT on leg tokens."
+    @event
+    ;; dupes checked in 'get-pair-create'
+    true)
+
+  (defcap ISSUING ()
+    "Private defcap for issuing operations."
+    true)
+
+  (defcap SWAPPING ()
+    "Private defcap for swapping operations."
+    true)
+
+  (defcap SWAP
+    ( sender:string
+      receiver:string
+      in:decimal
+      token-in:module{fungible-v2}
+      out:decimal
+      token-out:module{fungible-v2}
+    )
+    " Swap event debiting IN of TOKEN-IN from SENDER \
+    \ for OUT of TOKEN-OUT on RECEIVER."
+    @event
+    true
+  )
+
+  (defcap UPDATING ()
+    "Private defcap for updating operations."
+    true)
+
+  (defcap UPDATE
+    ( pair:string
+      reserve0:decimal
+      reserve1:decimal
+    )
+    "Event notifying reserves update for PAIR to RESERVE0 and RESERVE1."
+    @event
+    true
+  )
+
+  (defschema leg
+    token:module{fungible-v2}
+    reserve:decimal
+    )
+
+  (defschema pair
+    leg0:object{leg}
+    leg1:object{leg}
+    account:string
+    guard:guard
+    )
+
+  (deftable pairs:{pair})
+
+  (defconst MINIMUM_LIQUIDITY 0.1)
+
+  (defun init ()
+    (tokens.init-issuer (create-module-guard "issuance"))
+  )
+
+  (defun get-pair:object{pair}
+    ( tokenA:module{fungible-v2}
+      tokenB:module{fungible-v2}
+    )
+    (read pairs (get-pair-key tokenA tokenB)))
+
+  (defun pair-exists:bool
+    ( tokenA:module{fungible-v2}
+      tokenB:module{fungible-v2}
+    )
+    (with-default-read pairs
+      (get-pair-key tokenA tokenB)
+      { 'account: "" }
+      { 'account := a }
+      (> (length a) 0))
+  )
+
+  (defun enforce-deadline (deadline:time)
+    (enforce (>= deadline (at 'block-time (chain-data)))
+      "expired")
+  )
+
+  (defun update-reserves
+    ( p:object{pair}
+      pair-key:string
+      reserve0:decimal
+      reserve1:decimal
+    )
+    (require-capability (UPDATING))
+    (with-capability (UPDATE pair-key reserve0 reserve1)
+      (update pairs pair-key
+        { 'leg0: { 'token: (at 'token (at 'leg0 p))
+                 , 'reserve: reserve0 }
+        , 'leg1: { 'token: (at 'token (at 'leg1 p))
+                 , 'reserve: reserve1 }}))
+  )
+
+  (defun add-liquidity:object
+    ( tokenA:module{fungible-v2}
+      tokenB:module{fungible-v2}
+      amountADesired:decimal
+      amountBDesired:decimal
+      amountAMin:decimal
+      amountBMin:decimal
+      account-id:string
```

Comment:
This uses a single account name for both tokens.  Do we care that this prevents this function from working if the user doesn't have control of the same account on both tokens?

Reply:
User has to have the same account on both tokens, similarly to Uniswap.